File: Problems & Solutions/Z/Zero Sum Subarrays/22 December 2022/C#/Zero Sum Subarrays.cs
//{ Driver Code Starts
//Initial Template for C#


using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace DriverCode
{

    class GFG
    {
        static void Main(string[] args)
        {
            int testcases;// Taking testcase as input
            testcases = Convert.ToInt32(Console.ReadLine());
            while (testcases-- > 0)// Looping through all testcases
            {

                int n = Convert.ToInt32(Console.ReadLine());
                long[] arr = new long[n];
                var ip = Console.ReadLine().Trim().Split(' ');
                for (int i = 0; i < n; i++)
                {
                    arr[i] = long.Parse(ip[i]);
                }
                Solution obj = new Solution();
                var res = obj.findSubarray(arr, n);
                Console.WriteLine(res);
            }

        }
    }
}
// } Driver Code Ends

class Solution
{
    public long findSubarray(long[] arr, int n)
    {
        long sum = 0, counter = 0;

        Dictionary<long, long> mp = new Dictionary<long, long>();

        for (int i = 0; i < n; i++)
        {
            sum += arr[i];

            if (sum == 0)
            {
                counter++;
            }

            if (mp.ContainsKey(sum))
            {
                counter += mp[sum];
                mp[sum]++;
            } else {
                mp[sum] = 1;
            }

        }
        return counter;
    }

}

File: Problems & Solutions/Z/Zero Sum Subarrays/22 December 2022/Java/Zero Sum Subarrays.java
//{ Driver Code Starts
//Initial Template for Java

/*package whatever //do not write package name here */

import java.io.*;
import java.util.*;
class GFG {
	public static void main (String[] args) {
		Scanner sc=new Scanner(System.in);
		int t=sc.nextInt();
		sc.nextLine();
		while(t-->0)
		{
		    int n;
		    n=sc.nextInt();
		    
		    long arr[]=new long[n];
	
		    
		    for(int i=0;i<n;i++)
		    {
		        arr[i]=sc.nextLong();
		    }
            Solution ob = new Solution();
		    System.out.println(ob.findSubarray(arr,n));
		}
		
	}
}

// } Driver Code Ends

class Solution{
    public static long findSubarray(long[] arr ,long n) 
{
     long sum=0, counter=0; 
     HashMap<Long, Long>mp=new HashMap<>();
    
        for(int i=0;i<n;i++)
        {
            sum+=arr[i];
            
            if(sum==0)
            {
                counter++;
            }
            
            if(mp.containsKey(sum)) 
            {
               counter+=mp.get(sum);
            }
            
            if(mp.containsKey(sum))
            {
                long k=mp.get(sum);
                k++;
                mp.put(sum,k);
                
            }
            else
             mp.put(sum,1L);
        }
       
       return counter ;
}
}